subsys/bluetooth/audio/vocs_client.c

```diff
@@ -0,0 +1,688 @@
+		if (length == sizeof(inst->cli.location)) {
+			memcpy(&inst->cli.location, data, length);
+			BT_DBG("Location %u", inst->cli.location);
+		} else {
+			BT_DBG("Invalid length %u (expected %zu)",
+			       length, sizeof(inst->cli.location));
+			cb_err = BT_ATT_ERR_INVALID_ATTRIBUTE_LEN;
+		}
+	} else {
+		BT_DBG("Invalid location");
+		cb_err = BT_ATT_ERR_UNLIKELY;
+	}
+
+	if (inst->cli.cb && inst->cli.cb->location) {
+		inst->cli.cb->location(conn, inst, cb_err, inst->cli.location);
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static uint8_t internal_read_volume_offset_state_cb(struct bt_conn *conn, uint8_t err,
+						    struct bt_gatt_read_params *params,
+						    const void *data, uint16_t length)
+{
+	int cb_err = err;
+	struct bt_vocs *inst = lookup_vocs_by_handle(conn, params->single.handle);
+
+	memset(params, 0, sizeof(*params));
+
+	if (!inst) {
+		BT_ERR("Instance not found");
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (err) {
+		BT_WARN("Volume state read failed: %d", err);
+		cb_err = BT_ATT_ERR_UNLIKELY;
+	} else if (data) {
+		if (length == sizeof(inst->cli.state)) {
+			int write_err;
+
+			memcpy(&inst->cli.state, data, length);
+			BT_DBG("Offset %d, counter %u",
+			       inst->cli.state.offset,
+			       inst->cli.state.change_counter);
+
+			/* clear busy flag to reuse function */
+			inst->cli.busy = false;
+			write_err = bt_vocs_client_state_set(conn, inst, inst->cli.cp.offset);
+			if (write_err) {
+				cb_err = BT_ATT_ERR_UNLIKELY;
+			}
+		} else {
+			BT_DBG("Invalid length %u (expected %zu)", length, sizeof(inst->cli.state));
+			cb_err = BT_ATT_ERR_UNLIKELY;
+		}
+	} else {
+		BT_DBG("Invalid location");
+		cb_err = BT_ATT_ERR_UNLIKELY;
+	}
+
+	if (cb_err) {
+		inst->cli.busy = false;
+
+		if (inst->cli.cb && inst->cli.cb->set_offset) {
+			inst->cli.cb->set_offset(conn, inst, err);
+		}
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static void vcs_client_write_vocs_cp_cb(struct bt_conn *conn, uint8_t err,
+					struct bt_gatt_write_params *params)
+{
+	int cb_err = err;
+	struct bt_vocs *inst = lookup_vocs_by_handle(conn, params->handle);
+
+	memset(params, 0, sizeof(*params));
+
+	if (!inst) {
+		BT_DBG("Instance not found");
+		return;
+	}
+
+	BT_DBG("Inst %p: err: 0x%02X", inst, err);
+
+	/* If the change counter is out of data when a write was attempted from the application,
+	 * we automatically initiate a read to get the newest state and try again. Once the
+	 * change counter has been read, we restart the applications write request. If it fails
+	 * the second time, we return an error to the application.
+	 */
+	if (cb_err == BT_VOCS_ERR_INVALID_COUNTER && inst->cli.state_handle) {
+		BT_DBG("Invalid change counter. Reading volume state from server.");
+
+		inst->cli.read_params.func = internal_read_volume_offset_state_cb;
+		inst->cli.read_params.handle_count = 1;
+		inst->cli.read_params.single.handle = inst->cli.state_handle;
+
+		cb_err = bt_gatt_read(conn, &inst->cli.read_params);
+
+		if (cb_err) {
```

Comment:
So here the empty line really shouldn't be there. Assigning to a variable and checking for the variable's value is considered to be a single logical chunk of code.